### ISSUE TITLE
[localize] ICU select / plural / genders functions

### ISSUE THREAD

**Issue by aomarks** [lit/lit#1891]
Add support for ICU [select](https://unicode-org.github.io/icu-docs/apidoc/released/icu4j/com/ibm/icu/text/SelectFormat.html) and [plural](https://unicode-org.github.io/icu-docs/apidoc/released/icu4j/com/ibm/icu/text/PluralFormat.html) constructs, plus a helper for gender selects.

In extraction, we will analyze calls to these functions and emit the corresponding [ICU](https://unicode-org.github.io/icu/userguide/format_parse/messages/) syntax, e.g.:

```
{NUM_DRAFTS_1, plural,
  =0 {No drafts}
  =1 {1 draft}
  other {{NUM_DRAFTS_2} drafts}
}
```

When building for runtime mode, we'll emit matching function calls into the generated modules, with the localized templates substituted.

When building for transform mode, we can directly emit ternary expressions, or whichever expression ends up being the most concise, for each of the constructs, and elide the imports.

## select

Switch between N versions of a message. Typically used when a message varies by gender in the source locale.

```ts
import { select, other } from "@lit/localize/inflection.js";
 
select(this.targetGender, {
 female: html`Delete her account`,
 male: html`Delete his account`,
 [other]: html`Delete their account`,
});
```

#### Type signature

```ts
export const other = Symbol();
 
export declare function select<T extends string>(
 value: T,
 cases: { [X in T]?: HTMLTemplateResult } & {
   [other]: HTMLTemplateResult;
 }
);
```

## plural

Switch between N versions of a message, specifically for the case of pluralization. Differs from select because cases are limited to numbers, and because it encodes ICU plural semantics (which may be presented in a special way by translation tools).

```ts
import { plural, other } from "@lit/localize/inflection.js";
 
plural(this.numDrafts, {
 0: html`No drafts`,
 1: html`1 draft`,
 [other]: html`${this.numDrafts} drafts`,
});
```

#### Type signature

```ts
export const other = Symbol();
 
export declare function plural(
 num: number,
 cases: {
  [num: number]: HTMLTemplateResult;
  [other]: HTMLTemplateResult
 }
);
```

## genders

Generates translatable permutations of the given message for "male", "female", and "other" genders. Designed for the case where a message does not vary by gender in English, but may vary by gender in other languages. There can be up to 3 gender expressions, in which case all 27 permutations are generated.

Note this use case can also be represented with (possibly nested) selects on the 3 gender cases, where all English messages are duplicated. (In fact, the compiled ICU representation for genders is exactly this). But since this pattern is extremely common, the reduction in boilerplate seems worthwhile.

We also include this for feature parity with the [gender](https://github.com/google/closure-templates/blob/master/documentation/reference/messages.md#genders) tag in Closure templates.

```ts
import {genders} from '@lit/localize/inflection.js';
 
genders([this.guardianGender, this.childGender],
 html`Do you want to delete your child's account?`);
```

#### Type signature

```ts
export type Gender = "female" | "male" | "other";
 
export declare function genders(
 gender: [Gender] | [Gender, Gender] | [Gender, Gender, Gender],
 message: HTMLTemplateResult
);
```

**Comment by tirithen** [comment]
I have been using https://www.npmjs.com/package/intl-messageformat for this in several projects, it is part of https://formatjs.io/ and is a great package to take inspiration from if not use as is.

**Comment by marekbednar** [comment]
Any news on this please ? We are quite in a need of getting ICU on our translations for our large project. Thanks for a response in advance.

**Comment by aomarks** [comment]
No update on this unfortunately. It's still something I'd like to do, but it's not currently high on a priority list. Do you have any workarounds available?

**Comment by marekbednar** [comment]
> No update on this unfortunately. It's still something I'd like to do, but it's not currently high on a priority list. Do you have any workarounds available?

I'm trying to insert ICU string directly into both template and language file, such as:
${msg("{n, plural, =0 {keine Kommentare} one {1 Kommentar} other {# Kommentare}}", {id: 'feed_comments_label_plural'})}
I aim to make a wrapper over msg(), such as msgICU( ... decode icu and msg out)

**Comment by marekbednar** [comment]
using intl-messageformat it seems to work:
...
      <trans-unit xml:space="preserve" approved="no" id="feed_comments_label_plural"><source>{n, plural, =0 {keine Kommentare} one {1 Kommentar} other {# Kommentare}}</source>
        <target state="translated">{n, plural, =0 {keine Kommentare} one {1 Kommentar} other {# Kommentare}}</target>
      </trans-unit>
...


import {IntlMessageFormat} from "intl-messageformat"
.....
	msg(s: string, msgOptions: any, values: any): string | (string | void)[] | void {
		let translationByLocale = msg(s, msgOptions);
		return new IntlMessageFormat(translationByLocale, getLocale()).format(values);
	}
....

**Comment by pascalvos** [comment]
@marekbednar thank you for the idea that works indeed. depending on what your building this might be a good solution.

**Comment by gmzacharydovel** [comment]
Would you be interested in a PR on this?  Also, it would be really nice if we included the file location the localization was extracted from similar to po files.

**Comment by gmzacharydovel** [comment]
I'll probably have more time last week. I've already written wrappers for gettext so this probably will be similar.

**Comment by gmzacharydovel** [comment]
Quick question: Is there a reason lit doesn't support gettext?

**Comment by justinfagnani** [comment]
> Quick question: Is there a reason lit doesn't support gettext?

Only because localize was built to support some initial internal Google needs. It can grow to support more formats if there are contributors willing to maintain it. What would gettext support look like?

**Comment by gmzacharydovel** [comment]
Let me look into the codebase this week and make a proposal.  I'm trying to get my company on Lit and I may be able to shift us to prioritizing it.

My own older lib uses a plain text search to extract gettext strings.  If Lit is using AST (as I expect), I probably would adjust to that.

The flow wouldn't really change much.  I generally try to get non-engineers to use POEdit <https://poedit.net/> to translate stuff which supports both xliff and po files.

I haven't looked into Xliff that deeply yet to know what features are common between the two.

**Comment by gmzacharydovel** [comment]
This is the previous API I used for my gettext libraries:
https://github.com/zakkudo/translator

**Comment by sukima** [comment]
Most front-end projects I've seen all follow ICU message syntax through format.js

Might be worth considering supporting the same as it is very popular in this ecosystem.

Sadly gettext has fallen out of favor by the front-end industry as far as I can tell. Haven't seen gettext in use for over a decade now.

Food for thought.

**Comment by gmzacharydovel** [comment]
Yeah, I'm noticing that myself.  It is also following closer to the paradigm of ICU.  If I write a wrapper, it will most likely be around that.

I'm sorry for the delay on this.  I thought I would have some more free time, but I'll probably be clobbered with work until the end of next week.  For my next client I have interest in getting this implemented so it's coming regardless.

**Comment by gmzacharydovel** [comment]
> if been using https://formatjs.github.io/docs/intl-messageformat/ i think its idea is to build on https://402.ecma-international.org/ and i used MutationObserver of watch lang on the document to get the locale

I finally got some time to look a bit more into xliff.  I agree with you on using something 
like `intl-messageformat`, which should ideally be embedded in the `msg` method of `lit-localize`.

I'm going to hopefully play with this a bit more this next few weeks.... but it looks like the minimum effort might be to just update the documentation so that people can use their own messageformat provider.  Though this approach wouldn't support the prebuilt templates as imagined.

**Comment by gmzacharydovel** [comment]
Can this be assigned to me? Is there something I need to do?  It will help me to keep track of it
